System/IO/Temp.hs

```diff
@@ -212,3 +217,49 @@ mkPrivateDir s = System.Posix.createDirectory s 0o700
 -- "/home/feuerbach"
 getCanonicalTemporaryDirectory :: IO FilePath
 getCanonicalTemporaryDirectory = getTemporaryDirectory >>= canonicalizePath
+
+
+-- | Make a temporary file or directory name in the given directory, similar to unix 'mktemp'.
+-- This does not create a file.
+--
+-- Note that this might be subject to race conditions. Only use this on secure directories.
+mkTempFileName :: FilePath   -- ^ directory in which to test for available random filenames
+               -> String     -- ^ File name template. If the template is \"foo.ext\" then
+                             -- the created file will be \"fooXXX.ext\" where XXX is some
+                             -- random number. Note that this should not contain any path
+                             -- separator characters.
+               -> IO FilePath
+mkTempFileName tmp_dir template
+  | any isPathSeparator template
+  = failIO $ "mkTempFileName: Template string must not contain path separator characters: " ++ template
+  | otherwise = find' 0 (splitExtension template)
+ where
+  find' :: Int -> (String, String) -> IO FilePath
+  find' c (prefix, suffix)
+    | c >= max_tries = failIO "mkTempFileName: too many attempts of finding unique filename" 
+    | otherwise = do
+        rs <- rand_string
+        let filename = prefix ++ rs ++ suffix
+            filepath = tmp_dir </> filename
+        doesPathExist filepath >>= \case
+          False -> pure filepath
+          True -> find' (c + 1) (prefix, suffix)
+
+  max_tries = 10
+
+-- build large digit-alike number
+rand_string :: IO String
```

Comment:
This trick is directly from base `System.IO`.

Reply:
This "trick" is a large part why `mktemp` is not recommended for anything as it gives completely predictable filenames.

Reply:
Please file a bug against `base`.

Reply:
This is `temporary`, not `base`. `base` does a lot of awful things such as proliferation of String or the whole implicit text decoding/encoding business: I don't see how that makes it OK to put bad things in every other library too.

Regardless, as mentioned in the other comment, there's nothing wrong fundamentally with this function on its own but it can _only_ be used safely if the caller meets some about the usage.

Your caller does not do this.